dynamic_programming/catalan_numbers.cpp

```diff
@@ -1,75 +1,78 @@
-/** Print all the Catalan numbers from 0 to n, n being the user input.
-
- * A Catalan number satifies the following two properties:
+/**
+ * A Catalan number satisfies the following two properties:
  * C(0) = C(1) = 1; C(n) = sum(C(i).C(n-i-1)), from i = 0 to n-1
  * Read more about Catalan numbers here:
     https://en.wikipedia.org/wiki/Catalan_number
  */
 
-#include <iostream>
-using namespace std;
-
-int *cat;  // global array to hold catalan numbers
-
-unsigned long int catalan_dp(int n) {
-    /** Using the tabulation technique in dynamic programming,
-        this function computes the first `n+1` Catalan numbers
-
-        Parameter
-        ---------
-        n: The number of catalan numbers to be computed.
+#include <cassert>  /// for assert
+#include <cstdint>  /// for std::uint64_t
+#include <cstdlib>  /// for std::size_t
+#include <vector>   /// for std::vector
 
-        Returns
-        -------
-        cat[n]: An array containing the first `n+1` Catalan numbers
-    */
-
-    // By definition, the first two Catalan numbers are 1
-    cat[0] = cat[1] = 1;
-
-    // Compute the remaining numbers from index 2 to index n, using tabulation
-    for (int i = 2; i <= n; i++) {
-        cat[i] = 0;
-        for (int j = 0; j < i; j++)
-            cat[i] += cat[j] * cat[i - j - 1];  // applying the definition here
+/**
+ * @brief computes and caches Catalan numbers
+ */
+class catalan_numbers {
+    using value_type = std::uint64_t;
```

Comment:
Of course it is not.

But:
- for the user of this class it does not matter (note that it is even `private`),
- for the maintainer of this class it might be slightly easier: changing this type or _templating_ it will be easier.